picard/webservice/__init__.py

```diff
@@ -208,84 +353,61 @@ def _handle_reply(self, reply, request, handler, parse_response_type, refresh):
-                    else:
-                        log.error("Redirect loop: %s",
-                                  reply.request().url().toString(QUrl.RemoveUserInfo)
-                                  )
-                        handler(reply.readAll(), reply, error)
-                elif parse_response_type:
+                    self._handle_redirect(reply, request, redirect)
+                elif request.response_parser:
                     try:
-                        response_parser = self.get_response_parser(parse_response_type)
-                    except UnknownResponseParserError as e:
-                        log.error(e.args[0])
+                        document = request.response_parser(reply)
+                    except Exception as e:
+                        log.error("Unable to parse the response. %s", e)
                         document = reply.readAll()
-                    else:
-                        try:
-                            document = response_parser(reply)
-                        except Exception as e:
-                            log.error("Unable to parse the response. %s", e)
-                            document = reply.readAll()
                     finally:
                         handler(document, reply, error)
                 else:
                     handler(reply.readAll(), reply, error)
 
     def _process_reply(self, reply):
         try:
-            request, handler, parse_response_type, refresh = self._active_requests.pop(reply)
+            request = self._active_requests.pop(reply)
         except KeyError:
             log.error("Request not found for %s" % reply.request().url().toString(QUrl.RemoveUserInfo))
             return
         try:
-            self._handle_reply(reply, request, handler, parse_response_type, refresh)
+            self._handle_reply(reply, request)
         finally:
             reply.close()
             reply.deleteLater()
 
-    def get(self, host, port, path, handler, parse_response_type=DEFAULT_RESPONSE_PARSER_TYPE, priority=False,
-            important=False, mblogin=False, cacheloadcontrol=None, refresh=False, queryargs=None):
-        func = partial(self._start_request, "GET", host, port, path, None,
-                       handler, parse_response_type, mblogin, cacheloadcontrol=cacheloadcontrol, refresh=refresh, queryargs=queryargs)
-        return self.add_task(func, host, port, priority, important=important)
-
-    def post(self, host, port, path, data, handler, parse_response_type=DEFAULT_RESPONSE_PARSER_TYPE, priority=False, important=False, mblogin=True, queryargs=None):
+    def get(self, host, port, path, handler, parse_response_type=DEFAULT_RESPONSE_PARSER_TYPE,
+            priority=False, important=False, mblogin=False, cacheloadcontrol=None, refresh=False,
+            queryargs=None):
+        request = WSRequest("GET", host, port, path, handler, parse_response_type=parse_response_type,
```

Comment:
I wonder about creating WsGetRequest, WsPutRequest, etc... inheriting of a common WsRequest, it would perhaps simplify things a bit since we have many method-specific stuff in WsRequest. What do you think ?